tests/storage/linstor: Add test for forgetting and introducing Linstor SR

rushikeshjadhav opened the pull request:
Added `test_forget_and_introduce_sr in `test_linstor_sr.py` to validate forgetting and reintroducing an SR, ensuring correct PBD restoration and proper exception handling.
Updated SR class in sr.py with methods for retrieving parameters, getting type, forgetting, and introducing SR.

stormi commented:
I had a look, nothing I'd need to comment on I think, so I leave it to the other reviewers.

rushikeshjadhav commented on lib/host.py:
```diff
@@ -130,8 +130,14 @@ def stringify(key, value):
+        if isinstance(args, dict):
+            command += [stringify(key, value) for key, value in args.items()]
+        elif isinstance(args, list):
+            command += [stringify(key, value) for key, value in args]
```
Done

rushikeshjadhav commented on lib/sr.py:
```diff
@@ -147,6 +148,17 @@ def main_host(self):
+        return self.pool.master.xe('sr-param-get', {'uuid': self.uuid, 'param-name': param})
+
+    def type(self):
+        return self.pool.master.xe('sr-param-get', {'uuid': self.uuid, 'param-name': 'type'})
```
Thanks

rushikeshjadhav commented on lib/sr.py:
```diff
@@ -147,6 +148,17 @@ def main_host(self):
     def content_type(self):
         return self.pool.master.xe('sr-param-get', {'uuid': self.uuid, 'param-name': 'content-type'})
```
Done

rushikeshjadhav commented:
> @rushikeshjadhav did you get a chance to look at the ruff error?

I have addressed the ruff error and checked the test against my CI hosts. Looks good to me.

ydirson commented on lib/host.py:
```diff
@@ -130,8 +130,16 @@ def stringify(key, value):
+        # device-config:redundancy=2 device-config:provisioning=thin
+        command: List[str] = ['xe', action] + maybe_param_minimal + maybe_param_force
+        resolved_args = None
+        if isinstance(args, list):
```
Can you please git a full example of a commandline that would require this?

klmp200 committed:
Added test for forgetting and introducing Linstor SR `test_forget_and_introduce_sr`

- Validates the ability to forget an SR and properly reintroduce it.
- Ensures all PBD configurations are stored and restored correctly.
- Checks that exceptions are raised when querying a forgotten SR.
- Restores the SR and verifies successful reattachment of PBDs.
- Add `SR.introduce()` to introduce SRs in `sr.py`

Co-authored-by: Rushikesh Jadhav <rushikesh7@gmail.com>
Signed-off-by: Antoine Bartuccio <antoine.bartuccio@vates.tech>

glehmann commented on lib/host.py:
```diff
@@ -43,6 +43,7 @@
 XAPI_CONF_FILE = '/etc/xapi.conf'
```
_This comment was made on the commit metadata for commit e8df87b, on the Description line._

<!-- pulldash:commit-metadata sha=e8df87b69285f8012c164b657c9bcba2933c7417 line=9 label=Description -->

This is not a well formatted commit description. Is it missing the title?

glehmann replied:
_This comment was made on the commit metadata for commit 0a53601, on the Subject line._

<!-- pulldash:commit-metadata sha=0a53601e71c969343819c6abb0c8a3e88ee05bc5 line=8 label=Subject -->

This is missing a blank line after the title

glehmann replied:
_This comment was made on the commit metadata for commit 6485d5e, on the Subject line._

<!-- pulldash:commit-metadata sha=6485d5edd63308d40dc37d85dffbf9318700282f line=8 label=Subject -->

IMO you should squash that commit into the others

glehmann commented on lib/sr.py:
```diff
@@ -202,6 +202,12 @@ def param_clear(self, param_name: str) -> None:
     def content_type(self) -> str:
         return self.param_get('content-type')
 
+    def introduce(self, type: str, shared: bool, name_label: str, sr_uuid: str) -> str:
```
What is the return value? Could it be a `SR` object directly?

klmp200 replied:
That would definitely make sense

glehmann commented on lib/sr.py:
```diff
@@ -202,6 +202,12 @@ def param_clear(self, param_name: str) -> None:
     def content_type(self) -> str:
         return self.param_get('content-type')
 
+    def introduce(self, type: str, shared: bool, name_label: str, sr_uuid: str) -> str:
```
It doesn't seem connected to `self`, except to grab a pool master. IMO it should be a static method to highlight that in the method signature:

```suggestion
    @staticmethod
    def introduce(pool: Pool, type: str, shared: bool, name_label: str, sr_uuid: str) -> str:
```

Or it could be a pool method `introduce_sr(…) -> SR`

klmp200 replied:
I think it should be part of an SR because it introduces it.
It makes sense to be static as the SR doesn't really exist yet anyway.

glehmann reviewed (approved):
This still LGTM, and I don't see any new changes.
Am I missing some new things that would re-require a review?

klmp200 commented:
> This still LGTM, and I don't see any new changes. Am I missing some new things that would re-require a review?

Everything is good :)
